### ISSUE TITLE
Implement a reliable Action or command framework with the registry

### ISSUE THREAD

**Issue by tuan08** [DemandCube/NeverwinterDP_Legacy#564]
An action or a command may involve a long process with many steps. Each step should be reliably executed or reexecuted. The steps can be executed in the order or in parallel. For example the stop dataflow action can have the following steps:
1. Find out the the dataflow in the registry by the dataflow-id
2. Place or send an stop event in the dataflow events node
3. The dataflow master should listen to the dataflow event and pickup the stop event.
4. The dataflow master should find out all the dataflow worker, place a stop command in each dataflow worker
   1. The dataflow worker should catch the command and stop all the running dataflow task properly
   2. The dataflow master should wait for all the dataflow worker exit properly.
   3. The dataflow master should make the running dataflow status as failed and terminated
   4. The dataflow master should shutdown the backup dataflow master.
   5. The dataflow master shutdown itself
5. The scribengin master should detect a dataflow terminated and move it to the history

Create a framework that allow us to split an action into many steps. Those step should be store in the registry.  Each step should be reliably executed and mark as waiting, finished status... The master should be responsible to coordinate and schedule those steps. If the master crashed, the backup master suppose to be able to resume the steps.

**Comment by tuan08** [comment]
This framework has been implemented , but it will need to continue to review and update. The detail work will be entered in another task
